src/group_theory/exponent.lean

```diff
@@ -91,66 +112,185 @@ end
-  { apply exponent_min' _ _ h₁,
+  { apply exponent_min' _ h₁,
     simp_rw ←pow_eq_mod_exponent,
     exact hG },
   linarith,
 end
 
-variable [fintype G]
-
 @[to_additive lcm_add_order_of_dvd_exponent]
-lemma lcm_order_of_dvd_exponent : (finset.univ : finset G).lcm order_of ∣ exponent G :=
+lemma lcm_order_of_dvd_exponent [fintype G] : (finset.univ : finset G).lcm order_of ∣ exponent G :=
 begin
   apply finset.lcm_dvd,
   intros g hg,
-  exact order_dvd_exponent G g
+  exact order_dvd_exponent g
+end
+
+@[to_additive exists_order_of_eq_pow_padic_val_nat_add_exponent]
+lemma _root_.nat.prime.exists_order_of_eq_pow_padic_val_nat_exponent {p : ℕ} (hp : p.prime) :
+  ∃ g : G, order_of g = p ^ padic_val_nat p (exponent G) :=
+begin
+  haveI := fact.mk hp,
+  rcases (padic_val_nat p $ exponent G).eq_zero_or_pos with h | h,
+  { refine ⟨1, by rw [h, pow_zero, order_of_one]⟩ },
+  have he : 0 < exponent G := ne.bot_lt (λ ht, by {rw ht at h, exact h.ne' (padic_val_nat_zero _)}),
+  obtain ⟨g, hg⟩ : ∃ (g : G), g ^ (exponent G / p) ≠ 1,
+  { suffices key : ¬ exponent G ∣ exponent G / p,
+    { by simpa using mt (exponent_dvd_of_forall_pow_eq_one G (exponent G / p)) key },
+    exact λ hd, hp.one_lt.not_le ((mul_le_iff_le_one_left he).mp $
+                nat.le_of_dvd he $ nat.mul_dvd_of_dvd_div (dvd_of_one_le_padic_val_nat h) hd) },
+  obtain ⟨k, hk : exponent G = p ^ _ * k⟩ := pow_padic_val_nat_dvd; try {apply_instance},
+  obtain ⟨t, ht⟩ := nat.exists_eq_succ_of_ne_zero h.ne',
+  refine ⟨g ^ k, _⟩,
+  rw ht,
+  apply order_of_eq_prime_pow,
+  { rwa [hk, mul_comm, ht, pow_succ', ←mul_assoc, nat.mul_div_cancel _ hp.pos, pow_mul] at hg },
+  { rw [←nat.succ_eq_add_one, ←ht, ←pow_mul, mul_comm, ←hk],
+    exact pow_exponent_eq_one g },
 end
 
+end monoid
+
+section left_cancel_monoid
+
+variable [left_cancel_monoid G]
+
 @[to_additive lcm_add_order_eq_exponent]
-lemma lcm_order_eq_exponent {H : Type u} [fintype H] [left_cancel_monoid H] :
-  (finset.univ : finset H).lcm order_of = exponent H :=
+lemma lcm_order_eq_exponent [fintype G] : (finset.univ : finset G).lcm order_of = exponent G :=
 begin
-  apply nat.dvd_antisymm (lcm_order_of_dvd_exponent H),
-  apply exponent_dvd_of_forall_pow_eq_one,
-  intro g,
-  have h : (order_of g) ∣ (finset.univ : finset H).lcm order_of,
-  { apply finset.dvd_lcm,
-    exact finset.mem_univ g },
-  cases h with m hm,
+  apply nat.dvd_antisymm (lcm_order_of_dvd_exponent G),
+  refine exponent_dvd_of_forall_pow_eq_one G _ (λ g, _),
+  obtain ⟨m, hm⟩ : order_of g ∣ finset.univ.lcm order_of := finset.dvd_lcm (finset.mem_univ g),
   rw [hm, pow_mul, pow_order_of_eq_one, one_pow]
 end
 
+@[to_additive]
+lemma exponent_ne_zero_of_fintype [fintype G] : exponent G ≠ 0 :=
+by simpa [←lcm_order_eq_exponent, finset.lcm_eq_zero_iff] using λ x, (order_of_pos x).ne'
+
+end left_cancel_monoid
+
+section comm_monoid
+
+variable [cancel_comm_monoid G]
+
+@[to_additive] lemma exponent_eq_Sup_order_of (h : ∀ g : G, 0 < order_of g) :
```

Comment:
Do you see any way to split/golf this proof? Intuitively, I think it looks a lot longer than I would expect a proof of this result to be. But I haven't loaded it into VScode yet.

Reply:
my only thoughts are that the first half could maybe be made into a private lemma, and the second half could be turned into a `calc`; but I'm fairly bad at calc. If you want me to turn the first edge case into a separate lemma, though, I can do that!

Reply:
the order were a lot more finicky here than I expected; I wish I could just say "if `g` has order not dividing `t` (which maximises the group's order), then `g*t` clearly has larger order"; but this statement requires a fair bit more care than I thought